proposed/0006-Launcher.md

```diff
@@ -0,0 +1,192 @@
+
+**step 3:** Launcher checks the App-ID, reads the path from the `<LOCAL-CONFIG-FILE>` that it made and starts the app as an independent process. The Launcher supplies a random port on which it will listen to this app via command line options.
+
+**./path/to/XYZ --launcher “port:33000;protocol:udp;nonce:random-u32-number”**
+
+**step 4:** Launcher will wait for a predefined time of **15** seconds for data reception on that port. If it times out it will close the socket (release its binding to it)
+
+**step 5:** App responds on the socket asking for Launcher to give keys and its root directory, which Launcher had reserved as `XYZ-Root-Dir`
+- The payload format for this request shall be CBOR encoded sturcture of the following:
+```
+struct Request {
+    prefix: u8,  // This must be ASCII of '?' (a question mark)
+    nonce : u32, // This must be the nonce supplied by Launcher
+}
+```
+
+**step 6:** Launcher reads the `<LAUNCHER-CONFIG-FILE>` and creates a mapping in the MaidManagers for associating user’s `MAID Keys <-> App specific Keys` to allow App to PUT/POST on behalf of the user in `<APP-ROOT-DIR>`.
+- The mapping is done only for the ownership key/s (not encryption keys).
+- The request for mapping/un-mapping shall be a command to the `MaidManagers`. For commands to `MaidManagers`, `StructuredData` with special Type-Tag will be reserved. On reception of this `StructuredData` the vaults will check the payload and act upon the request instead of executing a normal reaction to the usual `PUT/DELETES`.
+- `StructuredData` with Tag-Type **9** will be reserved for `Client <-> Vault` messages.
+- The messages will be defined in [safe_vault](https://github.com/maidsafe/safe_vault).
+- This shall be the format of `Client <-> Vault` messages:
+```
+pub enum ClientVaultMessage {
+    MapKeys {
+        ownership_key_to_map   : sodiumoxide::crypto::sign::PublicKey, // Launcher-Assigned-Public-Key-For-App
+        ownership_key_to_map_to: sodiumoxide::crypto::sign::PublicKey, // User-PublicMaid-sign-key
+    },
+    UnMapKeys {
+        ownership_key_to_unmap     : sodiumoxide::crypto::sign::PublicKey, // Launcher-Assigned-Public-Key-For-App
+        ownership_key_to_unmap_from: sodiumoxide::crypto::sign::PublicKey, // User-PublicMaid-sign-key
+    },
+}
+```
+
+**step 7:** Launcher gives the **App Keys** and **App-Dir** to the App and at this point it closes its socket and is no longer associated with the App in anyway.
+- The payload format for this response shall be CBOR encoded sturcture of the following, followed by hybrid_encrypt of the stream using App-specific-crypto-keys:
+```
+struct Response {
+    root_directory_id    : NameType,
+    root_directory_tag   : u64
+    public_signing_key   : sodiumoxide::crypto::sign::PublicKey,
+    private_signing_key  : sodiumoxide::crypto::sign::SecretKey,
+    public_encrytion_key : sodiumoxide::crypto::box_::PublicKey,
+    private_encrytion_key: sodiumoxide::crypto::box_::SecretKey,
+}
+```
+
```

Comment:
Do you mean the App deterministically generates these keys ?

if yes then: I think randomly generated Maid keys are better and hence we use that stratergy in Login Packets - the Login Packet is encrypted with deterministic keys but the Maid Keys it contains which eventually is used to mutate the network is non-deterministic for every client and hence stored in Login Packet.

if not then: these will need to be stored somewhere